File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/AllTimetablesViewModel.java
package com.happs.ximand.ringcontrol.viewmodel.fragment;

import androidx.databinding.ObservableBoolean;
import androidx.databinding.ObservableInt;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.android.material.snackbar.Snackbar;
import com.happs.ximand.ringcontrol.FragmentNavigation;
import com.happs.ximand.ringcontrol.R;
import com.happs.ximand.ringcontrol.model.dao.BluetoothDao;
import com.happs.ximand.ringcontrol.model.dao.BluetoothEventListener;
import com.happs.ximand.ringcontrol.model.dao.SharedPreferencesDao;
import com.happs.ximand.ringcontrol.model.object.command.ReplaceTimetableCommand;
import com.happs.ximand.ringcontrol.model.object.command.simple.WeekendMode;
import com.happs.ximand.ringcontrol.model.object.info.BluetoothEvent;
import com.happs.ximand.ringcontrol.model.object.timetable.Lesson;
import com.happs.ximand.ringcontrol.model.object.timetable.Timetable;
import com.happs.ximand.ringcontrol.model.repository.impl.TimetableRepository;
import com.happs.ximand.ringcontrol.model.specification.impl.GetAllSqlSpecification;
import com.happs.ximand.ringcontrol.view.fragment.AddTimetableFragment;
import com.happs.ximand.ringcontrol.view.fragment.SettingsFragment;
import com.happs.ximand.ringcontrol.view.fragment.TimetableInfoFragment;
import com.happs.ximand.ringcontrol.viewmodel.dto.SnackbarDto;
import com.happs.ximand.ringcontrol.viewmodel.util.TimeUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class AllTimetablesViewModel extends BaseViewModel {

    private static final int APPLIED_TIMETABLE_NONE = -2;

    private final MutableLiveData<List<Timetable>> allTimetablesLiveData = new MutableLiveData<>();
    private final MutableLiveData<Integer> numOfTestLessonsLiveData = new MutableLiveData<>();
    private final BluetoothEventListener<BluetoothEvent> bluetoothEventListener;
    private final ObservableBoolean applyingPossible = new ObservableBoolean(false);
    private final ObservableInt lastAppliedTimetableId;

    private final MutableLiveData<Boolean> manualModeStateLiveData = new MutableLiveData<>();
    private final MutableLiveData<WeekendMode> weekendModeLiveData = new MutableLiveData<>();

    private boolean currentSendTimetableTask = false;
    private int lastSentTimetableId = APPLIED_TIMETABLE_NONE;

    public AllTimetablesViewModel() {
        this.numOfTestLessonsLiveData.setValue(2);
        this.bluetoothEventListener = new BluetoothEventListener<>(this::onBluetoothEvent);
        this.lastAppliedTimetableId = new ObservableInt(
                SharedPreferencesDao.getInstance().getAppliedTimetableId()
        );
        this.manualModeStateLiveData
                .setValue(SharedPreferencesDao.getInstance().getManualModeState());
        this.weekendModeLiveData.setValue(getCurrentWeekendMode());
        updateData();
        startConnectingToSavedDevice();
    }

    private WeekendMode getCurrentWeekendMode() {
        return WeekendMode.getInstanceForModeId(SharedPreferencesDao.getInstance().getWeekendMode());
    }

    public LiveData<List<Timetable>> getAllTimetablesLiveData() {
        return allTimetablesLiveData;
    }

    public LiveData<Integer> getNumOfTestLessonsLiveData() {
        return numOfTestLessonsLiveData;
    }

    public ObservableBoolean getApplyingPossible() {
        return applyingPossible;
    }

    public ObservableInt getLastAppliedTimetableId() {
        return lastAppliedTimetableId;
    }

    public LiveData<Boolean> getManualModeStateLiveData() {
        return manualModeStateLiveData;
    }

    public LiveData<WeekendMode> getWeekendModeLiveData() {
        return weekendModeLiveData;
    }

    public boolean isWeekendModeCaptionShouldBeShown(WeekendMode weekendMode) {
        int weekday = getCurrentWeekday();
        return !((weekday != Calendar.SATURDAY && weekday != Calendar.SUNDAY)
                || (weekday != Calendar.SUNDAY && weekendMode == WeekendMode.MODE_WORK_ON_SATURDAY)
                || (weekendMode == WeekendMode.MODE_WORK_ON_WEEKENDS));
    }

    private int getCurrentWeekday() {
        Calendar calendar = Calendar.getInstance();
        return calendar.get(Calendar.DAY_OF_WEEK);
    }

    private void onBluetoothEvent(BluetoothEvent event) {
        if (event == BluetoothEvent.READY) {
            applyingPossible.set(true);
            updateLastAppliedTimetableIfNecessary();
        } else if (applyingPossible.get()) {
            applyingPossible.set(false);
        }
    }

    private void updateLastAppliedTimetableIfNecessary() {
        if (currentSendTimetableTask && lastSentTimetableId != lastAppliedTimetableId.get()) {
            SharedPreferencesDao.getInstance()
                    .updateAppliedTimetableId(lastSentTimetableId);
        }
        currentSendTimetableTask = false;
    }

    @Override
    public boolean notifyOptionsMenuItemClicked(int itemId) {
        switch (itemId) {
            case R.id.toolbar_add:
                FragmentNavigation.getInstance().navigateTo(AddTimetableFragment.newInstance());
                return true;
            case R.id.toolbar_settings:
                FragmentNavigation.getInstance().navigateTo(SettingsFragment.newInstance());
                return true;
            default:
                return false;
        }
    }

    public void addTestLesson() {
        @SuppressWarnings("ConstantConditions")
        int numOfTestLessons = numOfTestLessonsLiveData.getValue();
        numOfTestLessonsLiveData.setValue(numOfTestLessons + 1);
    }

    public void removeTestLesson() {
        @SuppressWarnings("ConstantConditions")
        int numOfTestLessons = numOfTestLessonsLiveData.getValue();
        numOfTestLessonsLiveData.setValue(numOfTestLessons - 1);
    }

    public void updateData() {
        List<Timetable> timetables = TimetableRepository.getInstance()
                .query(new GetAllSqlSpecification());
        allTimetablesLiveData.setValue(timetables);
    }

    private void startConnectingToSavedDevice() {
        String savedAddress = SharedPreferencesDao.getInstance().getTargetDeviceAddress();
        BluetoothDao.getInstance().startConnectToDeviceTask(savedAddress);
    }

    public void showTimetableDetails(Timetable timetable) {
        TimetableInfoFragment fragment = TimetableInfoFragment.newInstance(timetable);
        FragmentNavigation.getInstance().navigateTo(fragment);
    }

    public void notifyAppliedTimetableUpdated() {
        SnackbarDto snackbarDto = new SnackbarDto(
                R.string.applied_timetable_updated, Snackbar.LENGTH_SHORT
        ).setActionResId(R.string.apply)
                .setActionClickListener(v -> applyUpdatedCurrentTimetable());
        getMakeSnackbarEvent().setValue(snackbarDto);
    }

    public void applyUpdatedCurrentTimetable() {
        int appliedTimetableId = SharedPreferencesDao.getInstance().getAppliedTimetableId();
        List<Timetable> allTimetables = allTimetablesLiveData.getValue();
        if (allTimetables != null) {
            applyTimetable(allTimetables.get(appliedTimetableId));
        }
    }

    public void applyTestTimetable() {
        @SuppressWarnings("ConstantConditions")
        int numOfTestLessons = numOfTestLessonsLiveData.getValue();
        List<Lesson> lessons = new ArrayList<>(numOfTestLessons * 2);
        for (int i = 1; i <= numOfTestLessons * 2; i += 2) {
            lessons.add(createLessonForTestTimetable(i));
        }
        applyTimetable(new Timetable(-1, "Test", lessons));
    }

    private Lesson createLessonForTestTimetable(int extraMinutes) {
        int number = extraMinutes / 2 + 1;
        return new Lesson(number,
                TimeUtils.getCurrentTimeWithFewMinutes(extraMinutes),
                TimeUtils.getCurrentTimeWithFewMinutes(extraMinutes + 1)
        );
    }

    public void applyTimetable(Timetable timetable) {
        byte[] command = new ReplaceTimetableCommand(timetable).getCommand();
        BluetoothDao.getInstance().sendMessage(command);
        currentSendTimetableTask = true;
        lastSentTimetableId = timetable.getId();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        BluetoothDao.getInstance().unsubscribeFromEvents(bluetoothEventListener);
    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/AllTimetablesViewModel.kt
package com.happs.ximand.ringcontrol.viewmodel.fragment

import android.bluetooth.BluetoothDevice
import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableInt
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.OnLifecycleEvent
import com.google.android.material.snackbar.Snackbar
import com.happs.ximand.ringcontrol.FragmentNavigation
import com.happs.ximand.ringcontrol.R
import com.happs.ximand.ringcontrol.model.`object`.command.ReplaceTimetableCommand
import com.happs.ximand.ringcontrol.model.`object`.command.simple.WeekendMode
import com.happs.ximand.ringcontrol.model.`object`.exception.BaseException
import com.happs.ximand.ringcontrol.model.`object`.timetable.Lesson
import com.happs.ximand.ringcontrol.model.`object`.timetable.Timetable
import com.happs.ximand.ringcontrol.model.bl.BluetoothCommunicator
import com.happs.ximand.ringcontrol.model.bl.callback.ConnectCallback
import com.happs.ximand.ringcontrol.model.bl.callback.SendCallback
import com.happs.ximand.ringcontrol.model.dao.SharedPreferencesDao
import com.happs.ximand.ringcontrol.model.repository.impl.TimetableRepository
import com.happs.ximand.ringcontrol.model.specification.impl.GetAllSqlSpecification
import com.happs.ximand.ringcontrol.view.fragment.AddTimetableFragment
import com.happs.ximand.ringcontrol.view.fragment.SettingsFragment
import com.happs.ximand.ringcontrol.view.fragment.TimetableInfoFragment.Companion.newInstance
import com.happs.ximand.ringcontrol.viewmodel.dto.SnackbarDto
import com.happs.ximand.ringcontrol.viewmodel.util.TimeUtils.getCurrentTimeWithFewMinutes
import java.util.*

class AllTimetablesViewModel : BaseViewModel(), SendCallback, ConnectCallback {

    val allTimetablesLiveData: MutableLiveData<MutableList<Timetable>> = MutableLiveData()
    val numOfTestLessonsLiveData = MutableLiveData<Int>()
    val applyingPossible = ObservableBoolean(false)
    val lastAppliedTimetableId: ObservableInt
    val manualModeStateLiveData = MutableLiveData<Boolean>()
    val weekendModeLiveData: MutableLiveData<WeekendMode> = MutableLiveData()

    private var currentSendTimetableTask = false
    private var lastSentTimetableId = APPLIED_TIMETABLE_NONE

    init {
        numOfTestLessonsLiveData.value = 2
        lastAppliedTimetableId = ObservableInt(
                SharedPreferencesDao.instance.getAppliedTimetableId()
        )
        manualModeStateLiveData.value = SharedPreferencesDao.instance.getManualModeState()
        weekendModeLiveData.value = getCurrentWeekendMode()
        updateData()
        startConnectingToSavedDevice()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        BluetoothCommunicator.instance.sendCallback = this
        BluetoothCommunicator.instance.connectCallback = this
    }

    private fun getCurrentWeekendMode(): WeekendMode {
        return WeekendMode.getInstanceForModeId(SharedPreferencesDao.instance.getWeekendMode())
    }

    fun isWeekendModeCaptionShouldBeShown(weekendMode: WeekendMode): Boolean {
        val weekday = getCurrentWeekday()
        return !(weekday != Calendar.SATURDAY && weekday != Calendar.SUNDAY
                || weekday != Calendar.SUNDAY && weekendMode == WeekendMode.MODE_WORK_ON_SATURDAY
                || weekendMode == WeekendMode.MODE_WORK_ON_WEEKENDS)
    }

    private fun getCurrentWeekday(): Int {
        val calendar = Calendar.getInstance()
        return calendar[Calendar.DAY_OF_WEEK]
    }

    override fun onConnected(device: BluetoothDevice) {
        applyingPossible.set(true)
    }

    override fun onSent(sentBytes: ByteArray) {
        if (sentBytes[0] == ReplaceTimetableCommand.CODE) {
            updateLastAppliedTimetableIfNecessary()
        }
    }

    override fun onException(e: BaseException) {
        applyingPossible.set(false)
    }

    private fun updateLastAppliedTimetableIfNecessary() {
        if (lastSentTimetableId != lastAppliedTimetableId.get()) {
            SharedPreferencesDao.instance
                    .updateAppliedTimetableId(lastSentTimetableId)
        }
        currentSendTimetableTask = false
    }

    override fun notifyOptionsMenuItemClicked(itemId: Int): Boolean {
        when (itemId) {
            R.id.toolbar_add -> {
                FragmentNavigation.instance.navigateTo(AddTimetableFragment.newInstance())
            }
            R.id.toolbar_settings -> {
                FragmentNavigation.instance.navigateTo(SettingsFragment.newInstance())
            }
            else -> return false
        }
        return true
    }

    fun addTestLesson() {
        val numOfTestLessons = numOfTestLessonsLiveData.value!!
        numOfTestLessonsLiveData.value = numOfTestLessons + 1
    }

    fun removeTestLesson() {
        val numOfTestLessons = numOfTestLessonsLiveData.value!!
        numOfTestLessonsLiveData.value = numOfTestLessons - 1
    }

    fun updateData() {
        allTimetablesLiveData.value = TimetableRepository.getInstance()
                .query(GetAllSqlSpecification())
    }

    private fun startConnectingToSavedDevice() {
        val savedAddress = SharedPreferencesDao.instance.getTargetDeviceAddress()
        BluetoothCommunicator.instance
                .connectToDevice(checkNotNull(savedAddress))
    }

    fun showTimetableDetails(timetable: Timetable?) {
        val fragment = newInstance(timetable)
        FragmentNavigation.instance.navigateTo(fragment)
    }

    fun notifyAppliedTimetableUpdated() {
        val snackbarDto = SnackbarDto(
                R.string.applied_timetable_updated, Snackbar.LENGTH_SHORT
        ).setActionResId(R.string.apply)
                .setActionClickListener { applyUpdatedCurrentTimetable() }
        makeSnackbarEvent.value = snackbarDto
    }

    private fun applyUpdatedCurrentTimetable() {
        val appliedTimetableId = SharedPreferencesDao.instance.getAppliedTimetableId()
        val allTimetables: List<Timetable>? = allTimetablesLiveData.value
        if (allTimetables != null) {
            applyTimetable(allTimetables[appliedTimetableId])
        }
    }

    fun applyTestTimetable() {
        val numOfTestLessons = numOfTestLessonsLiveData.value!!
        val lessons: MutableList<Lesson> = ArrayList(numOfTestLessons * 2)
        var i = 1
        while (i <= numOfTestLessons * 2) {
            lessons.add(createLessonForTestTimetable(i))
            i += 2
        }
        applyTimetable(Timetable(-1, "Test", lessons))
    }

    private fun createLessonForTestTimetable(extraMinutes: Int): Lesson {
        val number = extraMinutes / 2 + 1
        return Lesson(number,
                getCurrentTimeWithFewMinutes(extraMinutes),
                getCurrentTimeWithFewMinutes(extraMinutes + 1)
        )
    }

    fun applyTimetable(timetable: Timetable) {
        val command: ByteArray = ReplaceTimetableCommand(timetable).toByteArray()
        BluetoothCommunicator.instance.writeBytes(command)
        currentSendTimetableTask = true
        lastSentTimetableId = timetable.id
    }

    companion object {
        private const val APPLIED_TIMETABLE_NONE = -2
    }

}